homeassistant/components/metoffice/weather.py

```diff
@@ -1,127 +1,156 @@
+from homeassistant.components.weather import WeatherEntity
+from homeassistant.const import LENGTH_KILOMETERS, TEMP_CELSIUS
+from homeassistant.core import callback
+from homeassistant.helpers.typing import ConfigType, HomeAssistantType
+
+from .const import (
+    ATTRIBUTION,
+    CONDITION_CLASSES,
+    DEFAULT_NAME,
+    DOMAIN,
+    METOFFICE_COORDINATOR,
+    METOFFICE_DATA,
+    METOFFICE_NAME,
+    VISIBILITY_CLASSES,
+    VISIBILITY_DISTANCE_CLASSES,
 )
-from homeassistant.helpers import config_validation as cv
-
-from .sensor import ATTRIBUTION, CONDITION_CLASSES, MetOfficeCurrentData
 
 _LOGGER = logging.getLogger(__name__)
 
-DEFAULT_NAME = "Met Office"
-
-PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
-    {
-        vol.Required(CONF_API_KEY): cv.string,
-        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
-        vol.Inclusive(
-            CONF_LATITUDE, "coordinates", "Latitude and longitude must exist together"
-        ): cv.latitude,
-        vol.Inclusive(
-            CONF_LONGITUDE, "coordinates", "Latitude and longitude must exist together"
-        ): cv.longitude,
-    }
-)
-
-
-def setup_platform(hass, config, add_entities, discovery_info=None):
-    """Set up the Met Office weather platform."""
-    name = config.get(CONF_NAME)
-    datapoint = dp.connection(api_key=config.get(CONF_API_KEY))
 
-    latitude = config.get(CONF_LATITUDE, hass.config.latitude)
-    longitude = config.get(CONF_LONGITUDE, hass.config.longitude)
+async def async_setup_entry(
+    hass: HomeAssistantType, entry: ConfigType, async_add_entities
+) -> None:
+    """Set up the Met Office weather sensor platform."""
+    hass_data = hass.data[DOMAIN][entry.entry_id]
 
-    if None in (latitude, longitude):
-        _LOGGER.error("Latitude or longitude not set in Home Assistant config")
-        return
-
-    try:
-        site = datapoint.get_nearest_site(latitude=latitude, longitude=longitude)
-    except dp.exceptions.APIException as err:
-        _LOGGER.error("Received error from Met Office Datapoint: %s", err)
-        return
-
-    if not site:
-        _LOGGER.error("Unable to get nearest Met Office forecast site")
-        return
-
-    data = MetOfficeCurrentData(hass, datapoint, site)
-    try:
-        data.update()
-    except (ValueError, dp.exceptions.APIException) as err:
-        _LOGGER.error("Received error from Met Office Datapoint: %s", err)
-        return
-
-    add_entities([MetOfficeWeather(site, data, name)], True)
+    async_add_entities(
+        [MetOfficeWeather(entry.data, hass_data,)], False,
+    )
 
 
 class MetOfficeWeather(WeatherEntity):
     """Implementation of a Met Office weather condition."""
 
-    def __init__(self, site, data, name):
-        """Initialise the platform with a data instance and site."""
-        self._name = name
-        self.data = data
-        self.site = site
+    def __init__(self, entry_data, hass_data):
+        """Initialise the platform with a data instance."""
```

Comment:
Same as above, should the available property be set?

Reply:
Agreed. Done.